### PR TITLE
Provided python examples for graph creation and GUI interaction.

### PR DESCRIPTION
**Pull request by tpecar**
A set of examples which showcases the usage of libbabeltrace2 python bindings and how they might be used in conjunction with a GUI framework - PyQt5 is used in the examples.

Currently, the GUI examples are simply a table which displays events received by the sink
image
but the underlying architecture should make it a decent starting point for creating anything more complex.

---
There are multiple implementations of this same GUI example:
1. graph run in gui event loop (`can_graph_gui_event_loop.py`)
2. graph run in a thread, with events passed to the gui thread via signals (`can_graph_gui_threaded_simple.py`), which is thread safe, but not performant
3. graph run in a thread, with events being passed through a shared buffer (non-thread safe), with additional hacks to make the graph thread yield back execution to the gui thread (`can_graph_gui_threaded_responsive.py`)

The 1. option would be my recommended way of doing it.

Chronologically, the 2. option was the first one developed, 3. option is its evolution, and the 1. option was the last one implemented, when I got the hang of how Qt works.

### CODE REVIEW

**Comment by simark** [comment]
Could you please squash your commits into one, or a few which incrementally build up your work?

It's fine to have multiple commits in a pull request, but these commits should not represent the chronological progress that you've made while writing the code.  In other words, you shouldn't have a commit that introduces some error and a following commit that fixes it.  You should just amend the commit that introduced the error.  Instead, they should represent small / manageable steps that build towards the end goal of the series.

Thanks!

**Comment by tpecar** [comment]
Ok, I will squash the intermediate commits.

Just one question - do you find it useful that we keep all three different implementations of the gui example, or should I just keep the one that I find the best (`can_graph_gui_event_loop.py`)?

File: README.md [hunk 1]
```diff
@@ -1,16 +1,21 @@
 This repository contains a few silly or less silly Babetrace 2 plugins.
 
  * Very simple
-   [source and sink components](https://github.com/simark/babeltrace-fun-plugins/tree/master/my-first-components),
+   [source and sink components](/my-first-components),
```
Comment (simark) [inline README.md]: Just wondering, in which situation is that useful.

Could you submit these README fixes as its own PR?

**Comment by simark** [comment]
> Just one question - do you find it useful that we keep all three different implementations of the gui example, or should I just keep the one that I find the best (`can_graph_gui_event_loop.py`)?

I have just looked very quickly, but I think that if they are of increasing complexity, it can be useful to have these multiple examples.  The simple one can be useful to get people started, while the most complex one can help show a better design for an interactive application.

File: can/python/bt_plugin_can.py [hunk 2]
```diff
@@ -2,6 +2,13 @@
 import cantools
 import struct
 
+
+# It seems that in version 2.0.0, even though the framework supports creation of streams
+# without packet support, the utils.trimmer filter is not capable of parsing such streams.
```
Comment (simark) [inline can/python/bt_plugin_can.py:7]: Thanks for noticing this, I don't think it's the intention here, I'll investigate.

File: README.md (hunk 1 above)
Comment (tpecar) [reply README.md]: If the original and forked repository (or even a branch on the same repository) diverge significantly, it is better that the README links point to the files of its repository / branch.

I can submit them as a separate pull request, no problem.

**Comment by tpecar** [comment]
> I have just looked very quickly, but I think that if they are of increasing complexity, it can be useful to have these multiple examples. The simple one can be useful to get people started, while the most complex one can help show a better design for an interactive application.

Ok, the thing is that `can_graph_gui_threaded_*` examples really aren't the best way of doing things.
They essentially demonstrate that you don't really solve anything if you put graph execution in a separate thread - at least in Python. The results would probably be different if you did this directly in C/C++.

What I'll do is rework the `can_graph_gui_threaded_simple.py` to use the event loop approach, so that it can be used as an initial example.

**Comment by simark** [comment]
> Ok, the thing is that `can_graph_gui_threaded_*` examples really aren't the best way of doing things.
> They essentially demonstrate that you don't really solve anything if you put graph execution in a separate thread - at least in Python. The results would probably be different if you did this directly in C/C++.

Are you saying this because Python threads don't actually execute simultaneously?

**Comment by tpecar** [comment]
> Are you saying this because Python threads don't actually execute simultaneously?

As far as I understand, yes.

It does give you the benefit of having separate execution contexts which the python interpreter / os manage, but the problem is that you can't really control how these threads get scheduled.

In the threaded implementation I started to run in to the issue where the thread running the BT2 graph starved the main (GUI) thread, thereby making the GUI unresponsive. I did manage to solve this by periodically blocking the BT2 graph thread after a timed interval, but that just feels kludgy.

I'm aware that there's multiprocessing, but that opens a whole new world of issues of transferring data between processes that I didn't want to go into.

**Comment by tpecar** [comment]
Just FYI, I'm still working on this - currently in the middle of rewriting the advanced GUI example to better showcase the capabilities of the framework.

**Comment by tpecar** [comment]
Ok, I've rebased the commits according to your suggestions. I also tried to follow the commit header format that was used in the previous commits.
Original commits are available at [pull-request/2/before-rebase](https://github.com/tpecar/babeltrace-fun-plugins/tree/pull-request/2/before-rebase).

The examples were further cleaned up and improved. The examples use the graph.run_once() in Qt event loop, and that works reasonably well (whether one might call it fast is up for debate).

- the simple example `can_graph_gui_simple.py` is <100 loc and provides a basic live data view
  image
- the advanced example `can_graph_gui_advanced.py` now also parses field classes to give additional insight on data structures. It also demonstrates payload parsing, which is displayed in the tree view as "Last Value"
  image

Thread-based implementations were left in c98a6af, if one wishes to toy with that concept - they end up being marginally faster when the gui has to process a lot of events.

**Comment by simark** [comment]
Ok, thanks for doing this!  I have a lot on my plate these days, so I likely won't review it this week, but feel free to ping me if I haven't looked at it by ~2 weeks.

File: can/python/bt_plugin_can.py (hunk 2 above)
Comment (simark) [reply can/python/bt_plugin_can.py:7]: Can you try to remove this and test with the tip of the stable-2.0 branch of babeltrace?  If it works, then we can avoid creating packets for nothing here.  I'll ask when we plan to release babeltrace 2.0.2, which would include this fix.

File: can/python/bt_plugin_can.py [hunk 3]
```diff
@@ -19,7 +26,6 @@ def log_info(cur_log_level):
 def print_info(text):
     print("INFO: {}".format(text))
 
-
```
Comment (simark) [inline can/python/bt_plugin_can.py:22]: Can you format each of your commits with `black`?  We are using this in Babeltrace itself, and I'd like to use it here too.  It should re-add this removed empty line.

File: can/python/bt_plugin_can.py [hunk 4]
```diff
@@ -29,11 +35,47 @@ def __init__(self, config, port):
...
+        # Tuple:
+        # timestamp, frame_id, data
+        return struct.unpack("<ii8s", msg)
+
+    # Based on
+    #   tests/bindings/python/bt2/test_message_iterator.py
+    #
+    # TODO: check if timestamp from file actually in ns
+    #
+    def _user_seek_ns_from_origin(self, ns_from_origin):
```
Comment (simark) [inline can/python/bt_plugin_can.py:72]: Instead of implementing `_user_seek_ns_from_origin`, I'd suggest implementing `_user_seek_beginning`, which would be simpler (since you would simply do `self._file.seek(0)`).  If the iterator implements the `seek_beginning` method, Babeltrace's core does exactly what you are doing here (read messages until we reach the desired ns_from_origin).  Internally in the code base, we call this "auto seek": https://github.com/efficios/babeltrace/blob/master/src/lib/graph/iterator.c#L1156.  This helps provide a default (although not very efficient) implementation of `seek_ns_from_origin` with minimal effort.

Note that letting Babeltrace do this will be a bit less efficient than doing it yourself, since there will be more back and forths between C and Python.  However, I don't expect it to be significant.  If this ends up being a wrong assumption, I'll be happy to reconsider.

Implementing `_user_seek_ns_from_origin` is useful if the trace format has some kind of index which allows you to jump directly to the right place without having to scan all the events.

File: can/can_graph.py [hunk 5]
```diff
@@ -0,0 +1,294 @@
+#!/usr/bin/env python3
+# *_* coding: utf-8 *_*
+
+"""
+Creates and runs the same graph as the babeltrace-fun-plugins/can/check.sh utility.
+---
+Please note: libbabeltrace2 python library (bt2) depends on its core C library.
```
Comment (simark) [inline can/can_graph.py]: I would write this with the perspective that babeltrace will be `make install`'ed somewhere.  Running inside the build directory is kind of a corner case of babeltrace developers, whereas these examples are more meant for users.

So if babeltrace has been installed system-wide (e.g. in `/usr`), there is generally nothing to set.  But if it's been installed to some random location (e.g. in `/opt/babeltrace`), then you would need to set some variables.  Could you please adapt the explanations below to that scenario?

In my experience, it is necessary to set `LD_LIBRARY_PATH` and `PYTHONPATH`.  Did you also need to set `PYTHONPATH`?

So it would be:

    LD_LIBRARY_PATH=[babeltrace2 install directory]/lib
    PYTHONPATH=[babeltrace2 install directory]/lib/[python version]/site-packages

File: can/can_graph.py [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+    graph.run()
+
+# BT2 graph - trims existing CTF trace
+#
+# With the repo version of bt_plugin_gui, the utils.trimmer will fail with
+#   Babeltrace 2 library postcondition not satisfied;
+#   Packet message has no default clock snapshot:
+#
+# This seems to be a limitation of the 2.0.0. version of utils.trimmer filter.
+# You can work around it by extending the CANSource to generate a packet-enabled stream.
```
Comment (simark) [inline can/can_graph.py]: You should be able to remove this with Babeltrace 2.0.2.

File: can/can_graph.py [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+    # Load required components from plugins
+    sink = system_plugins['ctf'].sink_component_classes['fs']
+    filter = system_plugins['utils'].filter_component_classes['trimmer']
+    source = system_plugins['ctf'].source_component_classes['fs']
+
+    # Create graph and add components
+    graph = bt2.Graph()
+
+    graph_sink = graph.add_component(sink, 'test_sink',
+        params=bt2.MapValue({
```
Comment (simark) [inline can/can_graph.py]: I don't remember, is it actually needed to use `bt.MapValue`, or you can do simply:

```
params={
    'path': './ctf-filtered'
}
```

?  If possible, I'd use the more succinct version, keeping in mind that people are likely to just copy this.

File: can/can_graph.py [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+#   Babeltrace 2 library postcondition not satisfied;
+#   Packet message has no default clock snapshot:
+#
+# This seems to be a limitation of the 2.0.0. version of utils.trimmer filter.
+# You can work around it by extending the CANSource to generate a packet-enabled stream.
+#
+def graph_ctf_filter_ctf():
+
+    # Load required components from plugins
+    sink = system_plugins['ctf'].sink_component_classes['fs']
```
Comment (simark) [inline can/can_graph.py]: Can you name these variables `ctf_fs_sink_comp_cls`, `trimmer_filter_comp_cls` and `ctf_fs_source_comp_cls`.  This is just to make sure to show that these are component *classes*.

File: can/can_graph.py [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+    filter = system_plugins['utils'].filter_component_classes['trimmer']
+    source = system_plugins['ctf'].source_component_classes['fs']
+
+    # Create graph and add components
+    graph = bt2.Graph()
+
+    graph_sink = graph.add_component(sink, 'test_sink',
```
Comment (simark) [inline can/can_graph.py]: Likewise, can you name these variables `ctf_fs_sink_comp`, `trimmer_filter_comp` and `ctf_fs_source_comp`, to that that they are components?

File: can/can_graph.py [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+    # Run graph
+    graph.run()
+
+# BT2 graph - trims CANSource
+#
+# With the repo version of bt_plugin_gui, this will fail with
+#   Cannot make upstream message iterator initially seek
+#
+# You need to provide an implementation of
+#   def _user_seek_ns_from_origin(self, ns_from_origin):
```
Comment (simark) [inline can/can_graph.py]: Having to add either `_user_seek_ns_from_origin` or `_user_seek_beginning` is expected for the trimmer to work (since in the end it needs to be able to `seek_ns_from_origin`).  But since you did add this method to `CANIterator`, I'm not sure it's really useful to mention this?

File: can/can_graph.py [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+# in bt2._UserMessageIterator in order to make it work
+#
+def graph_can_filter_ctf():
```
Comment (simark) [inline can/can_graph.py]: The comments on `graph_ctf_filter_ctf` apply here too.

File: can/can_graph.py [hunk 12]
```diff
@@ -0,0 +1,294 @@
...
+    graph.connect_ports(
+        list(graph_source.output_ports.values())[0],
+        list(graph_sink.input_ports.values())[0]
+    )
+
+    # Run graph
+    graph.run()
+
+# BT2 graph - dumps CANSource trace to CTF
+def graph_can_ctf():
```
Comment (simark) [inline can/can_graph.py]: The comments on `graph_ctf_filter_ctf` apply here too.

File: can/can_graph.py (hunk 5 above)
Comment (simark) [reply can/can_graph.py]: Actually, maybe this info should go into the top-level `README.md`, since it will always be the same.

File: can/can_graph.py [hunk 13]
```diff
@@ -0,0 +1,294 @@
+#!/usr/bin/env python3
```
Comment (simark) [inline can/can_graph.py]: Can you modify `can/README.md` to explain how to run your examples?

In particular, the paths to the `python` directory seem to make it so you need to run the `can_graph.py` and `can_graph_gui_simple.py` examples while in the `can` directory, but you need to run the `can_graph_gui_responsive.py` example while in the `gui` directory.  If this could be made consistent, it would be nice.  But also, if it was documented in the README (with a command to copy paste) how to run each file, it would also help.

Comment (simark) [reply can/can_graph.py]: Oh, I just noticed that this all moves in the third and fourth commits.  Please remove the intermediate states from the PR.  It's not really helpful to do things a certain way only to move and change them in the following commit.  It's normal to do many trial and errors while designing, but for the PR, just keep the final one.

**Review by simark — CHANGES_REQUESTED** [review]

**Comment by tpecar** [comment]
Just to report in that I'm still around. Was busy with other projects, however I will find the time in the following week to incorporate the upstream changes + your suggestions.

**Comment by tpecar** [comment]
Hah, it's been a long week.

Apologies for going silent, I seem to have a talent for taking up more projects than what I can actually handle in a timely manner.

More so, the project for which this functionality was intended has stalled. It will be picked up again, but some things will need to be reworked.

I'll reopen / make a new pull request then (expect it in jan/feb next year).

Thank you for your input up to this point.
